src/catchup/AssumeStateWork.cpp

```diff
@@ -0,0 +1,59 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "AssumeStateWork.h"
+#include "bucket/BucketManager.h"
+#include "catchup/IndexBucketsWork.h"
+#include "work/WorkSequence.h"
+#include "work/WorkWithCallback.h"
+
+namespace stellar
+{
+AssumeStateWork::AssumeStateWork(Application& app,
+                                 HistoryArchiveState const& has,
+                                 uint32_t maxProtocolVersion)
+    : Work(app, "assume-state", BasicWork::RETRY_NEVER)
+    , mHas(has)
+    , mMaxProtocolVersion(maxProtocolVersion)
+{
+}
+
+BasicWork::State
+AssumeStateWork::doWork()
+{
+    if (!mWorkSpawned)
+    {
+        std::vector<std::shared_ptr<BasicWork>> seq;
+
+        mApp.getBucketManager().assumeState(mHas);
```

Comment:
I'm worried that here we assign partially valid (unindexed) buckets to the bucketlist. If something further down goes wrong, then core is left in an invalid state. Safer would be to generate indexes to buckets first, then assign fully valid buckets to the bucketlist. The thing is that this cleanup only matters when the flag is on, so we could do this in a subsequent change

Reply:
This is a good point, but will take a fair bit of work since `Buckets` referenced by `HistoryArchiveState` are not yet managed by `BucketManager`. Commenting here for visibility, will address in follow-up PR.

Reply:
yeah, it might be annoying to change (you'd need to pass around a list of bucket pointers not attached to the bucketlist instead I think). Best is to open an issue, so we can track this.